### DISCUSSION TITLE
Lesson 7 Part 18 | Cheaper Withdraw Continued : Complete Array as memory

### DISCUSSION THREAD

**Discussion by devrajsinghrawat** [Cyfrin/foundry-full-course-cu#628] [Q&A] [answered]
Hi Folks,

I am re writing withdraw function as suggested in below video, which gave me gas optimisation as 

```
    function Cheaperwithdraw() external onlyOwner {
        uint256 m_funderLength = s_funders.length;

        for (uint256 funderIndex=0; funderIndex < m_funderLength; funderIndex++){
            address funder = s_funders[funderIndex];
            s_addressToAmountFunded[funder] = 0;
        }
        s_funders = new address[](0);
        // call
        (bool callSuccess, ) = payable(msg.sender).call{value: address(this).balance}("");
        require(callSuccess, "Call failed");
    } 
```
https://youtu.be/ST_4j4vsadk?list=PL2-Nvp2Kn0FPH2xU3IbKrrkae-VVXs1vk&t=203

**Original withdraw:** `FundMeTest:testOwnerCanWithdrawMultiFunder() (gas: 512343)`

**As suggested in video:** ` FundMeTest:testOwnerCanCheaperWithdrawMultiFunder() (gas: 511565)`

But instead if I go one step further and do it in below way then it is even more optimised 
-  Here I have declared  m_funders as a array which is copied into memory once and later used every place when s_funders should be used.

```
    function Cheaperwithdraw() external onlyOwner {
        address[] memory m_funders = s_funders;
        uint256 m_funderLength = m_funders.length;

        for (uint256 funderIndex=0; funderIndex < m_funderLength; funderIndex++){
            address funder = m_funders[funderIndex];
            s_addressToAmountFunded[funder] = 0;
        }
        s_funders = new address[](0);
        // call
        (bool callSuccess, ) = payable(msg.sender).call{value: address(this).balance}("");
        require(callSuccess, "Call failed");
    }
```

**Results:** ` FundMeTest:testOwnerCanCheaperWithdrawMultiFunder() (gas: 511445)`

I would to know your suggestion on this if this is the correct approach? 

2) when I am  doing this  `address[] memory m_funders = s_funders; ` that time is it coping the whole array into memory or just taking ref from storage?

**Comment by theirrationalone** [comment, marked as answer]
Answer 1: In the Context of Gas Optimization..... Yes Of course it's more efficient, same trick that Patrick(SIR) used in his `freecodecamp` Hardhat course.

Answer2: No each time it is creating a new array. It's a major difference in solidity and other programming langs.

This chunk 👎below... creates a new array into memory(means it Pushes a new array into stack & pops it out after function execution).
```solidity
address[] memory m_funders = s_funders;
```

that's why we haven't reseted that `m_funcders` array.

```solidity
 s_funders = new address[](0);
```

Suggestion to Gas Optimization: Try to make shorter Names of  `functions`, `variables`, `constants`, `arrays`, almost of everything... It also helps to optimize Gas. Similar kind of approach but way optimized but for testing purposes is often used and that's called **`SYMBOLIC EXECUTION`**.

Caution ⚠: Be aware, flood bunch of testssss into it to check if any vulnerability it Poses??? You can do **STATEFUL INVARIANT FUZZ TESTING**.

Suggestion to Test: Try to find out is it clearing all the mappings `(funder -> amount)` or not?? ....
